src/db.rs

```diff
@@ -27,30 +29,123 @@ use std::collections::BTreeMap;
 use std::ffi::{CStr, CString};
 use std::fmt;
 use std::fs;
+use std::marker::PhantomData;
 use std::path::Path;
 use std::path::PathBuf;
 use std::ptr;
 use std::slice;
 use std::str;
+use std::sync::RwLock;
 use std::time::Duration;
 
+// Marker trait to specify single or multi threaded column family alterations for DB
+// Also, this is minimum common API sharable between SingleThreaded and
+// MultiThreaded. Others differ in self mutability and return type.
+pub trait ThreadMode {
+    fn new(cf_map: BTreeMap<String, ColumnFamily>) -> Self;
+    unsafe fn cf_drop_all(&mut self);
+}
+
+pub struct SingleThreaded {
+    map: BTreeMap<String, ColumnFamily>,
+}
+
+pub struct MultiThreaded {
+    map: RwLock<BTreeMap<String, ColumnFamily>>,
+}
+
+impl ThreadMode for SingleThreaded {
+    fn new(cf_map: BTreeMap<String, ColumnFamily>) -> Self {
+        Self { map: cf_map }
+    }
+
+    unsafe fn cf_drop_all(&mut self) {
+        for cf in self.map.values() {
+            ffi::rocksdb_column_family_handle_destroy(cf.inner);
+        }
+    }
+}
+
+impl ThreadMode for MultiThreaded {
+    fn new(cf_map: BTreeMap<String, ColumnFamily>) -> Self {
+        Self {
+            map: RwLock::new(cf_map),
+        }
+    }
+
+    unsafe fn cf_drop_all(&mut self) {
+        for cf in self.map.read().unwrap().values() {
+            ffi::rocksdb_column_family_handle_destroy(cf.inner);
+        }
+    }
+}
+
 /// A RocksDB database.
 ///
 /// See crate level documentation for a simple usage example.
-pub struct DB {
+pub struct DbWithThreadMode<T: ThreadMode> {
     pub(crate) inner: *mut ffi::rocksdb_t,
-    cfs: BTreeMap<String, ColumnFamily>,
+    cfs: T, // Column families are held differently depending on thread mode
     path: PathBuf,
 }
 
+pub trait InternalDbAdapter {
+    fn inner(&self) -> *mut ffi::rocksdb_t;
+
+    fn get_opt<K: AsRef<[u8]>>(
+        &self,
+        key: K,
+        readopts: &ReadOptions,
+    ) -> Result<Option<Vec<u8>>, Error>;
+
+    fn get_cf_opt<K: AsRef<[u8]>>(
+        &self,
+        cf: impl ColumnFamilyRef,
+        key: K,
+        readopts: &ReadOptions,
+    ) -> Result<Option<Vec<u8>>, Error>;
+}
+
+impl<T: ThreadMode> InternalDbAdapter for DbWithThreadMode<T> {
+    fn inner(&self) -> *mut ffi::rocksdb_t {
+        self.inner
+    }
+
+    fn get_opt<K: AsRef<[u8]>>(
+        &self,
+        key: K,
+        readopts: &ReadOptions,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        self.get_opt(key, readopts)
+    }
+
+    fn get_cf_opt<K: AsRef<[u8]>>(
+        &self,
+        cf: impl ColumnFamilyRef,
+        key: K,
+        readopts: &ReadOptions,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        self.get_cf_opt(cf, key, readopts)
+    }
+}
+
+/// DB struct for single-threaded column family creations/deletions
+/// Note: Previously this was direct struct; type-aliased for compatibility. Directly
+/// use DbWithThreadMode<MultiThreaded> for multi-threaded column family alternations.
```

Comment:
as a matter of fact, I tried to be explicit to avoid confusion that SingleThreaded is inherently slow because of some kind of being single-threaded. (it's just that column alternations is single-threaded, still rocksdb are usually multi-threaded unless configured specifically).

Reply:
maybe, should I add a `note:` comment like this to be more explicit, or too verbose for general library audience?

```rust
// Note: Even with SingleThreaded, almost all of RocksDB operations is multi-threaded
// unless RocksDB is specifically configured otherwise. SingleThreaded only forces serialization of
// column family alternations by requring `&mut self` of DB instance due to its implementation details.
// MultiThreaded can be appropriate for the situation of multi-threaded workloads where the workload
// even includes column family alternations, costing the RwLock overhead inside DB.
```

Reply:
done! 2cfe746ae0dbe5a40c92c0e61f925bddd1f59714

I also sectionized the explanations.